src/main/java/com/databricks/jdbc/dbclient/impl/http/HttpRequestTypeBasedRetryHandler.java

```
@@ -0,0 +1,140 @@
+      int retryDelayMillis = RetryHandlingHelperFunctions.calculateExponentialBackoff(attempt);
+      retryDelayMillis = RetryHandlingHelperFunctions.addJitter(retryDelayMillis);
+      try {
+        CloseableHttpResponse response = httpClient.execute(request);
+        int statusCode = response.getStatusLine().getStatusCode();
+
+        // Get retry delay from strategy
+        retryDelayMillis = strategy.retryRequestAfter(response, attempt, connectionContext);
+        if (retryDelayMillis == -1) {
+          return response; // Strategy says don't retry
+        }
+        // apply jitter if strategy says to retry
+        retryDelayMillis = RetryHandlingHelperFunctions.addJitter(retryDelayMillis);
+
+        switch (statusCode) {
+          case HttpStatus.SC_SERVICE_UNAVAILABLE:
+            tempUnavailableTimeout -= retryDelayMillis;
+          case HttpStatus.SC_GATEWAY_TIMEOUT:
+            rateLimitTimeout -= retryDelayMillis;
+        }
+
+        // Check whether the connection context allows to wait until next attempt
+        if (tempUnavailableTimeout <= 0 || rateLimitTimeout <= 0) {
+          LOGGER.debug(
+              "Retry timeout exceeded for {} on attempt {}, received HTTP status code {}. Returning response",
+              requestType,
+              attempt,
+              statusCode);
+          return response;
+        }
+
+        // Last attempt check
+        if (attempt > maxRetries) {
+          LOGGER.debug(
+              "Max retries ({}) reached for {} on attempt {}, returning response",
+              maxRetries,
+              requestType,
+              attempt);
+          return response;
+        }
+
+        String errorReason = response.getStatusLine().getReasonPhrase();
+        String errorMessage =
+            String.format(
+                "Retry failure. HTTP response code: %s, Error Message: %s",
+                statusCode, errorReason);
+        LOGGER.debug(errorMessage);
```

don't get this part, we just log a silent error?

---

Changed to Logger.Error(errorMessage)